packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -240,52 +242,57 @@ export class DefaultCompletionProvider implements CompletionProvider {
+    protected fillCompletionItem(document: TextDocument, offset: number, item: CompletionValueItem): CompletionItem | undefined {
         let label: string;
-        if (typeof value === 'string') {
-            label = value;
-        } else if (isAstNode(value)) {
-            const name = this.nameProvider.getName(value);
+        if (typeof item.label === 'string') {
+            label = item.label;
+        } else if (isAstNode(item.label)) {
+            const name = this.nameProvider.getName(item.label);
             if (!name) {
                 return undefined;
             }
             label = name;
-        } else if (!isAstNode(value)) {
-            label = value.name;
+        } else if (item.label && !isAstNode(item.label)) {
+            label = item.label.name;
         } else {
             return undefined;
         }
-        const textEdit = this.buildCompletionTextEdit(document, offset, label);
+        let insertText: string;
+        if (typeof item.textEdit?.newText === 'string') {
+            insertText = item.textEdit.newText;
+        } else if (typeof item.insertText === 'string') {
+            insertText = item.insertText;
+        } else {
+            insertText = label;
+        }
+        const textEdit = this.buildCompletionTextEdit(document, offset, label, insertText);
         if (!textEdit) {
             return undefined;
         }
-        const item: CompletionItem = { label, textEdit, ...info };
-        return item;
+        const completionItem: CompletionItem = { textEdit, ...item, label };
+        return completionItem;
     }
 
-    protected buildCompletionTextEdit(document: TextDocument, offset: number, completion: string): TextEdit | undefined {
-        let negativeOffset = 0;
+    protected buildCompletionTextEdit(document: TextDocument, offset: number, label: string, newText: string): TextEdit | undefined {
         const content = document.getText();
-        const contentLowerCase = content.toLowerCase();
-        const completionLowerCase = completion.toLowerCase();
-        for (let i = completionLowerCase.length; i > 0; i--) {
-            const contentLowerCaseSub = contentLowerCase.substring(offset - i, offset);
-            if (completionLowerCase.startsWith(contentLowerCaseSub) && (i === 0 || !this.isWordCharacterAt(contentLowerCase, offset - i - 1))) {
-                negativeOffset = i;
-                break;
-            }
-        }
-        if (negativeOffset > 0 || offset === 0 || !this.isWordCharacterAt(completion, 0) || !this.isWordCharacterAt(content, offset - 1)) {
-            const start = document.positionAt(offset - negativeOffset);
+        const tokenStart = this.backtrackToTokenStart(content, offset);
+        const identifier = content.substring(tokenStart, offset);
+        if (this.charactersFuzzyMatch(identifier, label.toLowerCase())) {
```

Comment:
Proposal: move the `toLowerCase()` call into the `charactersFuzzyMatch` method to support when someone would like to change the behavior to case-sensitive exact match.